### ISSUE TITLE
Can we add obfuscated gradient attack (BPDA)? 

### ISSUE THREAD

**Issue by Shawn-Shan** [tensorflow/cleverhans#1050]
Since BPDA breaks everything, we should include it to help defense paper to evaluate against it?

**Comment by npapernot** [comment]
Feel free to start a new pull request in one of the following languages: tf1/2/pytorch/jax

**Comment by carlini** [comment]
BPDA is less of an actual attack itself and more of a strategy that can be used with any existing attack. Given almost any gradient-based attack, you can use the BPDA idea to get better gradients for a defense that causes problems with gradient descent.

It would probably be easier to just use pre-existing attacks, but pass augmented models that do the BPDA trick when asked for their gradient, than made multiple copies of each attack.

**Comment by yaoshiang** [comment]
The paper you wrote w/ Athalya and Wagner doesn't detail how to approximate a layer f^I with an approximator g. Would you be willing to describe it? Was it simply a dense layer with linear activation (linear approximation), or, perhaps a weighted average of an ensemble of dense layers with common activations (linear/tanh/relu/sigmoid/leakyrelu), with the weights set using exponential moving average?

**Comment by carlini** [comment]
There is no one approximation that works, unfortunately. Instead, for any given defense, you'll have to think about what is the best function that is (a) nicely differentiable, but also (b) a close approximation of the original function. Designing such functions can be difficult and is time consuming.
